BIODIVERSITY IN NATIONAL PARKS - P. Brister 

BI DATA ANALYST PORTFOLIO PROJECT 

Introduction 

This is Data Science Portfolio project about Biodiversity in four National Parks in the United States of America (Bryce National Park, Great Smokey Mountain National Park, Yellow Stone National Park, and Yosemite National Park), that explores data about species, their classification, conservation status and the number of observations of the different species on each park. 

Here are a few questions that this project has sought to answer:
    •What is the distribution of conservation status for species?
    •Are certain types of species more likely to be endangered? 
    •Which species is more prevalent and what is their distribution amongst the parks?
    •Which species is the least prevalent amongst each national park? 
    
    
 

For this project, the inital material consits of two datasets: 'Observations' and 'Species' - in csv format and provided by Codecademy.com. 

Note: The data for this project is inspired by real data, but is mostly fictional. 

Scoping 

Four sections were created below to help guide the project's process and progress. The first section is the project goals, this section will define the high level objectives and set the intentions for this project. The next section is the data, luckily in this project, the data is already provided but still needs to be checked of project goals can be with the available data. Thirdly, the analysis will have to be throught through, which include methods and questions that are aligned with the project goals. Lastly, evaluation will help us build conclusions and findings from our analysis. 


Project Goals 

In this project the perspective will be through a biodiversity analyst for the National Parks Service. The National Parks service wants to ensure the survival of at-risk species, to maintain the level of biodiversity within thier parks. Therefore, the main objectives as an analyst will be understanding characteristics about the species and ther conservation status, and those species and thier relationship to the national parks. Some questions posed: 

  •What is the distribution of conservation status for species?
  •Are certain types of species more likely to be endangered? 
  •Which species is more prevalent and what is their distribution amongst the parks?
  •Which species is the least prevalent amongst each national park? 
  

Data 

This project has two datasets that came with the package. The first csv file has information about each species and another has observations of species with park locations. This data will be used to analyze the goal of this project. 

Analysis

In this section, descriptive analysis and data visualization techniques will be employed to understand the data better. 



Evaluation

Lastly, it's a good idea to revisit the goals and check if the output of the analysis corresponds to the questions first set to be answered. This section will also reflect on what has been learned through the process and if any of th questions were able or unable to be answered. This could also include limitations or if any of the analysis could have been done sing different methodologies. 

IMPORT Python Modules

In [1]:
#Import Modules 
import pandas as pd 
import numpy as py 
import seaborn as sns 
import matplotlib.pyplot as plt 


LOADING Data

Loading data into DataFrames - so that the data can be explored and visualized.  

In [2]:
#Open and create dataframe 
species_info = pd.read_csv('species_info.csv')
species_info.head(10)

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
4   Mammal                 Cervus elaphus   
5   Mammal         Odocoileus virginianus   
6   Mammal                     Sus scrofa   
7   Mammal                  Canis latrans   
8   Mammal                    Canis lupus   
9   Mammal                    Canis rufus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)                 NaN  
4                                      Wapiti Or Elk                 NaN  
5                                  White-Tailed Deer                 NaN  
6                                Feral Hog, Wild Pig                 NaN  
7                                             Coyote  Species of Concern  
8                                          Gray Wolf          Endangered  
9                                           Red Wolf          Endangered

species

The species_info.csv file contains information on the different species in the National Parks. The columns in the dataset include: 
    •Category - the category of Taxonomy for each species 
    •Scientific name - the scientific name of each species 
    •Common names - the common name of each species (important to note that there are multiple entries in one row)
    •Conservation status - the species conservation status 

In [3]:
observations = pd.read_csv('observations.csv')
observations.head(10)

scientific_name                            park_name  \
0                 Vicia benghalensis  Great Smoky Mountains National Park   
1                     Neovison vison  Great Smoky Mountains National Park   
2                  Prunus subcordata               Yosemite National Park   
3               Abutilon theophrasti                  Bryce National Park   
4           Githopsis specularioides  Great Smoky Mountains National Park   
5  Elymus virginicus var. virginicus               Yosemite National Park   
6                   Spizella pusilla            Yellowstone National Park   
7                  Elymus multisetus  Great Smoky Mountains National Park   
8             Lysimachia quadrifolia               Yosemite National Park   
9         Diphyscium cumberlandianum            Yellowstone National Park   

   observations  
0            68  
1            77  
2           138  
3            84  
4            85  
5           112  
6           228  
7            39  
8           168  
9           250

observations 

The observations.csv file contains information from recorded sightings of different species throughout the national parks in the past 7 days. The columns included are: 
    •scientific name - the scientific name of each species 
    •park name - the name of the national park 
    •observations - the number of observations in the past 7 days 


Data characteristics 

check for the dimensions of the datasets' 

In [4]:
species_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


There are 5824 rows and 4 columns in the species.csv file. Notice - there are only 191 entries in the conservation_status column. Will investigate further. 

In [5]:
observations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23296 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  23296 non-null  object
 1   park_name        23296 non-null  object
 2   observations     23296 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 546.1+ KB


There are 23926 rows and 3 columns in the observations.csv file.

EDA - Exploratory Data Analysis 

Let's explore the species data a little more in depth. 

In [6]:
species_info.describe()

category    scientific_name        common_names  \
count             5824               5824                5824   
unique               7               5541                5504   
top     Vascular Plant  Castor canadensis  Brachythecium Moss   
freq              4470                  3                   7   

       conservation_status  
count                  191  
unique                   4  
top     Species of Concern  
freq                   161

There are 5824 entries in the category column. There are 7 unique categories, the Vascular Plant is the most frequent at 4470. 

There are 5824 entries in the scientific name column. There are 5541 scientific names and the Castor Canadensis is the most frequent at 3. 

There are 5824 entries in the common names column. There are 5504 common names and Brachythecium Moss is the most frequent at 7. I also mentioned earlier, that there seem to be mutliple entires in the column name and we need to clear that up before we merge dataframes. 

There are 191 entires in the conservation_status column. There are 4 unique categories and Species of Concern is the most frequent at 161. How many NAN values are there? 

In [7]:
#Clean column_names 
species_info['common_names'] = species_info['common_names'].str.split(',').str[0]
species_info.head(10)

category                scientific_name              common_names  \
0   Mammal  Clethrionomys gapperi gapperi  Gapper's Red-Backed Vole   
1   Mammal                      Bos bison            American Bison   
2   Mammal                     Bos taurus                   Aurochs   
3   Mammal                     Ovis aries            Domestic Sheep   
4   Mammal                 Cervus elaphus             Wapiti Or Elk   
5   Mammal         Odocoileus virginianus         White-Tailed Deer   
6   Mammal                     Sus scrofa                 Feral Hog   
7   Mammal                  Canis latrans                    Coyote   
8   Mammal                    Canis lupus                 Gray Wolf   
9   Mammal                    Canis rufus                  Red Wolf   

  conservation_status  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
5                 NaN  
6                 NaN  
7  Species of Concern  
8          Endangered  
9          Endangered

In [8]:
species_info.isnull().sum()

category                  0
scientific_name           0
common_names              0
conservation_status    5633
dtype: int64

In [9]:
species_info.conservation_status.unique()

array([nan, 'Species of Concern', 'Endangered', 'Threatened',
       'In Recovery'], dtype=object)

There are 4 different categories in the conservation status column and their entries are: 
•Species of Concern
•Endangered 
•In Recovery 
•Threatened 

There are also 5633 entries that do not have a value. Normally NAN values must be treated carefully - however, the absense of this suggest that these values are missing because they are not of concern. We are going to replace NAN with Least Concern. 

In [10]:
species_info.fillna('Least Concern', inplace=True)
species_info.groupby('conservation_status').size()

conservation_status
Endangered              16
In Recovery              4
Least Concern         5633
Species of Concern     161
Threatened              10
dtype: int64

All the NAN values have been replaced with Least Concern. We now have a better outlook on this column. Currently there are 16 entries that are at risk of being endangered, 4 recovering from previously being endangered, 161 that are of concern, 10 that are at risk of extinction, and lastly 5633 that are of least concern. 

In [11]:
#identify and delete duplicates
species_info.duplicated().sum()

species_info.drop_duplicates(inplace=True)

In [12]:
species_info.drop_duplicates(inplace=True)

There were 72 duplicates in the species_info dataset. They have been removed. 

In [13]:
species_info.describe()

category scientific_name common_names conservation_status
count             5752            5752         5752                5752
unique               7            5541         5297                   5
top     Vascular Plant   Puma concolor      A Sedge       Least Concern
freq              4426               3           10                5569

In [28]:
species_info.groupby('category').size()

category
Amphibian              79
Bird                  507
Fish                  127
Mammal                201
Nonvascular Plant     333
Reptile                79
Vascular Plant       4426
dtype: int64

Earlier, we identified that there were 7 unique classifications in the category column and Vascular Plant was the most frequent. Now, we have a deeper insight into the unique classifications and how many species were counted in each category. 
  

EDA - Exploratory Data Analysis pt2. 

Let's explore the observations data a little more in depth. 

In [15]:
observations.park_name.unique()

array(['Great Smoky Mountains National Park', 'Yosemite National Park',
       'Bryce National Park', 'Yellowstone National Park'], dtype=object)

In [16]:
observations.groupby('park_name').sum('observations')

observations
park_name                                        
Bryce National Park                        576025
Great Smoky Mountains National Park        431820
Yellowstone National Park                 1443562
Yosemite National Park                     863332

The dataset consist of data from 4 national parks: 
•Byrce National Park 
•Great Smokey Mountains National Park 
•Yellowstone National Park 
•Yosemite National Park 

In the past 7 days, there have been 3,314,739 observations recorded from all national parks. 

In [17]:
print(f'There have been {observations.observations.sum()} recorded over the last 7 days.')

There have been 3314739 recorded over the last 7 days.


In [18]:
observations.isnull().sum()

scientific_name    0
park_name          0
observations       0
dtype: int64

In [29]:
print(f'There are {observations.duplicated().sum()} duplicates in this DataFrame.')


There are 0 duplicates in this DataFrame.


In [30]:
#Update existing dataframe by removing duplicates 
observations.drop_duplicates(inplace=True)

There were no missing values in the dataset. However, there were 15 duplicates. They have been removed.

Now we are going to merge the DataFrames together. 


In [31]:
merged_data = observations.merge(species_info, on='scientific_name', how='left')
merged_data.head(10)

scientific_name                            park_name  \
0                 Vicia benghalensis  Great Smoky Mountains National Park   
1                     Neovison vison  Great Smoky Mountains National Park   
2                  Prunus subcordata               Yosemite National Park   
3               Abutilon theophrasti                  Bryce National Park   
4           Githopsis specularioides  Great Smoky Mountains National Park   
5  Elymus virginicus var. virginicus               Yosemite National Park   
6                   Spizella pusilla            Yellowstone National Park   
7                  Elymus multisetus  Great Smoky Mountains National Park   
8             Lysimachia quadrifolia               Yosemite National Park   
9         Diphyscium cumberlandianum            Yellowstone National Park   

   observations           category                common_names  \
0            68     Vascular Plant                Purple Vetch   
1            77             Mammal               American Mink   
2           138     Vascular Plant                Klamath Plum   
3            84     Vascular Plant                  Velvetleaf   
4            85     Vascular Plant              Common Bluecup   
5           112     Vascular Plant            Virginia Wildrye   
6           228               Bird               Field Sparrow   
7            39     Vascular Plant            Big Squirreltail   
8           168     Vascular Plant         Whorled Loosestrife   
9           250  Nonvascular Plant  Cumberland Diphyscium Moss   

  conservation_status  
0       Least Concern  
1       Least Concern  
2       Least Concern  
3       Least Concern  
4       Least Concern  
5       Least Concern  
6       Least Concern  
7       Least Concern  
8       Least Concern  
9       Least Concern

In [32]:
merged_data.isnull().sum()

scientific_name        0
park_name              0
observations           0
category               0
common_names           0
conservation_status    0
dtype: int64

After merging the DataFrames, it was important to ensure that there were no missing values.

Now that the two datasets have been merged, lets continue to clean it up beofre we dive into out analysis. I've decided to rename the park_name to National Park and remove National Park from each entry for cleaner look. 

In [38]:
merged_data = merged_data.rename(columns ={'park_name': 'National_Park'})
merged_data.head(10)

scientific_name                        National_Park  \
0                 Vicia benghalensis  Great Smoky Mountains National Park   
1                     Neovison vison  Great Smoky Mountains National Park   
2                  Prunus subcordata               Yosemite National Park   
3               Abutilon theophrasti                  Bryce National Park   
4           Githopsis specularioides  Great Smoky Mountains National Park   
5  Elymus virginicus var. virginicus               Yosemite National Park   
6                   Spizella pusilla            Yellowstone National Park   
7                  Elymus multisetus  Great Smoky Mountains National Park   
8             Lysimachia quadrifolia               Yosemite National Park   
9         Diphyscium cumberlandianum            Yellowstone National Park   

   observations           category                common_names  \
0            68     Vascular Plant                Purple Vetch   
1            77             Mammal               American Mink   
2           138     Vascular Plant                Klamath Plum   
3            84     Vascular Plant                  Velvetleaf   
4            85     Vascular Plant              Common Bluecup   
5           112     Vascular Plant            Virginia Wildrye   
6           228               Bird               Field Sparrow   
7            39     Vascular Plant            Big Squirreltail   
8           168     Vascular Plant         Whorled Loosestrife   
9           250  Nonvascular Plant  Cumberland Diphyscium Moss   

  conservation_status  
0       Least Concern  
1       Least Concern  
2       Least Concern  
3       Least Concern  
4       Least Concern  
5       Least Concern  
6       Least Concern  
7       Least Concern  
8       Least Concern  
9       Least Concern

In [39]:
merged_data['National_Park'] = merged_data['National_Park'].apply(lambda x: x.replace(' National Park', ''))
merged_data.head(10)

scientific_name          National_Park  observations  \
0                 Vicia benghalensis  Great Smoky Mountains            68   
1                     Neovison vison  Great Smoky Mountains            77   
2                  Prunus subcordata               Yosemite           138   
3               Abutilon theophrasti                  Bryce            84   
4           Githopsis specularioides  Great Smoky Mountains            85   
5  Elymus virginicus var. virginicus               Yosemite           112   
6                   Spizella pusilla            Yellowstone           228   
7                  Elymus multisetus  Great Smoky Mountains            39   
8             Lysimachia quadrifolia               Yosemite           168   
9         Diphyscium cumberlandianum            Yellowstone           250   

            category                common_names conservation_status  
0     Vascular Plant                Purple Vetch       Least Concern  
1             Mammal               American Mink       Least Concern  
2     Vascular Plant                Klamath Plum       Least Concern  
3     Vascular Plant                  Velvetleaf       Least Concern  
4     Vascular Plant              Common Bluecup       Least Concern  
5     Vascular Plant            Virginia Wildrye       Least Concern  
6               Bird               Field Sparrow       Least Concern  
7     Vascular Plant            Big Squirreltail       Least Concern  
8     Vascular Plant         Whorled Loosestrife       Least Concern  
9  Nonvascular Plant  Cumberland Diphyscium Moss       Least Concern

In [41]:
merged_data = merged_data.rename(columns={'common_names': 'common_name'})
merged_data.head()

scientific_name          National_Park  observations  \
0        Vicia benghalensis  Great Smoky Mountains            68   
1            Neovison vison  Great Smoky Mountains            77   
2         Prunus subcordata               Yosemite           138   
3      Abutilon theophrasti                  Bryce            84   
4  Githopsis specularioides  Great Smoky Mountains            85   

         category     common_name conservation_status  
0  Vascular Plant    Purple Vetch       Least Concern  
1          Mammal   American Mink       Least Concern  
2  Vascular Plant    Klamath Plum       Least Concern  
3  Vascular Plant      Velvetleaf       Least Concern  
4  Vascular Plant  Common Bluecup       Least Concern

In [44]:
merged_data.describ

scientific_name National_Park  observations        category  \
count            25006         25006  25006.000000           25006   
unique            5541             4           NaN               7   
top      Puma concolor         Bryce           NaN  Vascular Plant   
freq                36          6257           NaN           19182   
mean               NaN           NaN    142.272934             NaN   
std                NaN           NaN     69.922932             NaN   
min                NaN           NaN      9.000000             NaN   
25%                NaN           NaN     86.000000             NaN   
50%                NaN           NaN    124.000000             NaN   
75%                NaN           NaN    195.000000             NaN   
max                NaN           NaN    321.000000             NaN   

       common_name conservation_status  
count        25006               25006  
unique        5297                   5  
top        A Sedge       Least Concern  
freq            40               24198  
mean           NaN                 NaN  
std            NaN                 NaN  
min            NaN                 NaN  
25%            NaN                 NaN  
50%            NaN                 NaN  
75%            NaN                 NaN  
max            NaN                 NaN

Analysis 


5. Conclusions